conda_pypi/main.py

```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+            if path.name == "direct_url.json" and path.parent.suffix == ".dist-info":
+                data = json.loads(path.read_text())
+                if data.get("dir_info", {}).get("editable"):
+                    ignore = True
+                    break
+            if checksum and path.name == "RECORD" and path.parent.suffix == ".dist-info":
+                hashed_record = compute_record_sum(path, checksum)
+            if path.name == "WHEEL" and path.parent.suffix == ".dist-info":
+                for line in path.read_text().splitlines():
+                    line = line.strip()
+                    if ":" not in line:
+                        continue
+                    key, value = line.split(":", 1)
+                    if key == "Tag":
+                        wheel.setdefault(key, []).append(value.strip())
+                    else:
+                        wheel[key] = value.strip()
+        if ignore:
+            continue
+        if record.url:
+            lines.append(f"# pypi: {record.url}")
+        else:
+            seen = {"abi": set(), "platform": set()}
+            lines.append(f"# pypi: {record.name}=={record.version}")
+            lines[-1] += f" --python-version {python_details['version']}"
+            lines[-1] += f" --implementation {python_details['implementation']}"
+            if wheel and (wheel_tag := wheel.get("Tag")):
+                for tag in wheel_tag:
+                    _, abi_tag, platform_tag = tag.split("-", 2)
+                    if abi_tag != "none" and abi_tag not in seen["abi"]:
+                        lines[-1] += f" --abi {abi_tag}"
+                        seen["abi"].add(abi_tag)
+                    if platform_tag != "any" and platform_tag not in seen["platform"]:
+                        lines[-1] += f" --platform {platform_tag}"
+                        seen["platform"].add(platform_tag)
+            # Here we could try to run a --dry-run --report some.json to get the resolved URL
+            # but it's not guaranteed we get the exact same source so for now we defer to install
+            # time
+        if checksum and hashed_record:
+            lines[-1] += f" -- --record-checksum={checksum}:{hashed_record}"
+
+    return lines
+
+
+def dry_run_pip_json(args: Iterable[str], force_reinstall: bool = False) -> dict[str, Any]:
+    # pip can output to stdout via `--report -` (dash), but this
+    # creates issues on Windows due to undecodable characters on some
+    # project descriptions (e.g. charset-normalizer, amusingly), which
+    # makes pip crash internally. Probably a bug on their end.
+    # So we use a temporary file instead to work with bytes.
+    json_output = NamedTemporaryFile(suffix=".json", delete=False)
+    json_output.close()  # Prevent access errors on Windows
+
+    cmd = [
+        sys.executable,
+        "-mpip",
+        "install",
+        "--dry-run",
+        "--ignore-installed",
+        *(("--force-reinstall",) if force_reinstall else ()),
```

Comment:
what's with the unpacking?

```suggestion
        "--force-reinstall" if force_reinstall else "",
```

Reply:
I don't want to pass an empty string which subprocess will quote as a positional arg, won't it?

Reply:
Changed to conditional extensions here in general.